File: modules/drive-sense/android/src/main/java/expo/modules/drivesense/TimeBase.kt
package expo.modules.drivesense

import android.os.SystemClock
import kotlin.math.abs

/**
 * Time base and second windows — a port of `src/extract/timebase.ts` (README §7 "Time base",
 * "Windows"). Sensor events and fixes are stamped on the boot clock (`elapsedRealtimeNanos`) and
 * converted through ONE anchor per capture, so a wall-clock change mid-drive cannot reorder them.
 */
data class ClockAnchor(val epochMs: Double, val clockMs: Double) {
  companion object {
    /** Both clocks read back to back (README §7): `currentTimeMillis` and `elapsedRealtimeNanos / 1e6`. */
    fun now(): ClockAnchor =
      ClockAnchor(System.currentTimeMillis().toDouble(), SystemClock.elapsedRealtimeNanos() / 1e6)
  }
}

object TimeBase {
  data class Converted(val t: Double, val fellBack: Boolean)

  /**
   * A boot-clock instant on the capture's epoch base: `anchor.epochMs + (clockMs − anchor.clockMs)`.
   * Android measures an item's ARRIVAL with this (the boot clock at delivery), not with
   * `currentTimeMillis()` (README §7, review N2N3 I2): row `ts` is on the same base, so a wall-clock
   * step mid-drive cannot trip the sanity fallback and push samples and fixes off their windows.
   */
  fun anchoredNow(anchor: ClockAnchor, clockNowMs: Double): Double = anchor.epochMs + (clockNowMs - anchor.clockMs)

  /** [anchoredNow] at this instant. */
  fun anchoredNow(anchor: ClockAnchor): Double = anchoredNow(anchor, SystemClock.elapsedRealtimeNanos() / 1e6)

  /** Boot-clock ms → epoch ms; more than TIMEBASE_MAX_SKEW_MS from the arrival time falls back to it. */
  fun toEpochMs(clockMs: Double, anchor: ClockAnchor, arrivalEpochMs: Double): Converted {
    val t = anchor.epochMs + (clockMs - anchor.clockMs)
    return if (abs(t - arrivalEpochMs) > TIMEBASE_MAX_SKEW_MS) Converted(arrivalEpochMs, true) else Converted(t, false)
  }

  /** Exclusive start of the window closing at [ts]. */
  fun windowStart(prevTs: Long?, ts: Long): Long =
    if (prevTs == null || ts - prevTs > MAX_ROW_GAP_MS) ts - FIRST_WINDOW_MS else prevTs

  data class Window<T>(val inWindow: List<T>, val rest: MutableList<T>, val dropped: Int)

  /**
   * Split [buffer] (any order) into `(start, ts]` sorted oldest first (stable, like the reference's
   * `Array.prototype.sort`), what waits for later windows (`t > ts`), and a count of what belonged to
   * an already closed window.
   */
  fun <T> takeWindow(buffer: List<T>, start: Long, ts: Long, timeOf: (T) -> Double): Window<T> {
    val inWindow = ArrayList<T>()
    val rest = ArrayList<T>()
    var dropped = 0
    val s = start.toDouble()
    val e = ts.toDouble()
    for (item in buffer) {
      val t = timeOf(item)
      if (t > e) rest.add(item) else if (t > s) inWindow.add(item) else dropped++
    }
    return Window(inWindow.sortedBy(timeOf), rest, dropped)
  }

  /** The window's fix: the latest fix timestamp in `(start, ts]`, whatever the arrival order. */
  fun pickFix(fixes: List<FixSample>, start: Long, ts: Long): FixSample? {
    var best: FixSample? = null
    for (f in fixes) {
      if (f.t > start && f.t <= ts && (best == null || f.t > best.t)) best = f
    }
    return best
  }
}

File: modules/drive-sense/src/extract/timebase.ts
// Time base and second windows (README §7 "Time base" and "Windows"; review I2).
//
// Sensor and fix timestamps come on a monotonic boot clock (Android `elapsedRealtimeNanos`, iOS
// `systemUptime`). They are converted to epoch ms through ONE anchor taken when the capture
// starts, so a wall-clock change mid-drive cannot reorder samples. Each row's window is
// (previous row's ts, this ts] — never (ts − 1000, ts] off a jittery timer, which would drop or
// double-count samples — except for the first row of a capture or after a stall.
import type { FixSample } from './types';

/** A converted timestamp further than this from its arrival time falls back to the arrival time. */
export const TIMEBASE_MAX_SKEW_MS = 2000;
/** A gap between row timestamps longer than this starts over: the next window is (ts − 1000, ts]. */
export const MAX_ROW_GAP_MS = 2000;
/** The window of a capture's first row (and of the first row after a stall). */
export const FIRST_WINDOW_MS = 1000;

/**
 * Taken once per capture, at `startCapture`, reading both clocks back to back:
 * Android `{ System.currentTimeMillis(), SystemClock.elapsedRealtimeNanos() / 1e6 }`;
 * iOS `{ Date().timeIntervalSince1970 * 1000, ProcessInfo.processInfo.systemUptime * 1000 }`.
 */
export interface ClockAnchor {
  epochMs: number;
  clockMs: number;
}

/**
 * Boot-clock ms → epoch ms. `arrivalEpochMs` is the wall clock when the sample reached native code;
 * a conversion more than TIMEBASE_MAX_SKEW_MS away from it (a device with a different sensor time
 * base) falls back to it, and `fellBack` lets the port count it for diagnostics.
 */
export function toEpochMs(
  clockMs: number,
  anchor: ClockAnchor,
  arrivalEpochMs: number
): { t: number; fellBack: boolean } {
  const t = anchor.epochMs + (clockMs - anchor.clockMs);
  return Math.abs(t - arrivalEpochMs) > TIMEBASE_MAX_SKEW_MS
    ? { t: arrivalEpochMs, fellBack: true }
    : { t, fellBack: false };
}

/** Exclusive start of the window closing at `ts`. */
export function windowStart(prevTs: number | null, ts: number): number {
  return prevTs === null || ts - prevTs > MAX_ROW_GAP_MS ? ts - FIRST_WINDOW_MS : prevTs;
}

/**
 * Split a buffer (any order) into the window `(start, ts]`, sorted oldest first, and what is kept
 * for later windows (`t > ts`). Items at or before `start` belong to a closed window and are
 * dropped; `dropped` counts them.
 */
export function takeWindow<T extends { t: number }>(
  buffer: readonly T[],
  start: number,
  ts: number
): { inWindow: T[]; rest: T[]; dropped: number } {
  const inWindow: T[] = [];
  const rest: T[] = [];
  let dropped = 0;
  for (const item of buffer) {
    if (item.t > ts) rest.push(item);
    else if (item.t > start) inWindow.push(item);
    else dropped++;
  }
  inWindow.sort((a, b) => a.t - b.t);
  return { inWindow, rest, dropped };
}

/** The window's fix: the one with the latest fix timestamp in `(start, ts]`, whatever the arrival order. */
export function pickFix(fixes: readonly FixSample[], start: number, ts: number): FixSample | null {
  let best: FixSample | null = null;
  for (const f of fixes) {
    if (f.t > start && f.t <= ts && (best === null || f.t > best.t)) best = f;
  }
  return best;
}
